### DISCUSSION TITLE
What exactly is the reason not to use H2 in production?

### DISCUSSION THREAD

**Discussion by flowt-au** [keycloak/keycloak#16266] [Q&A] [answered]
Hi. I am fairly new to Java and very new to KeyCloak. While I am learning KC I have it running in container with a docker volume for the H2 database. This allows me to play with configuration and customisation without losing my data! ;-)

So, my question is: given I will be running KeyCloak in a container in production (eventually) and I am trying to minimise my resource requirements, and my use-case will have only one KC instance and maybe a few hundred users, what are the actual disadvantages of just using H2?

Thanks for a terrific product,
Murray

**Comment by xgp** [comment, marked as answer]
I think it depends on the size of your installation. If you only ever intend to have one node, a small user base (="a few hundred"), and you are backing up your docker volume frequently, it is just fine. I have worked with many companies using it in this exact way. I have heard second-hand some stories of database corruption, but I have yet to experience it myself.

**Reply by flowt-au** [reply]
Thanks Garth. :-)

**Reply by xgp** [reply]
@flowt-au just as a follow up, I posted some code I have used before that does periodic H2 runtime backups by scheduling a task with the Keycloak TimerProvider. Normally, H2 has to be stopped to use the offline backup tool, but this allows it to do it while Keycloak is running. So, if you are running an installation with a single node, you can mount the H2 data directory to a volume that you backup regularly (must support snapshots), and mount another directory which will contain the backup zip files produced by this extension. 

https://github.com/xgp/keycloak-experiments/tree/main/src/main/java/com/github/xgp/keycloak/backup

A couple of additional notes:
- You'll need to set the interval to something sane, like hourly or daily, as I think it's currently set to 1 minute.
- I haven't built in any logic to trim the directory where the backup files get saved, but that needs to be taken care of.
- This hasn't been heavily tested in a production environment, so I would review and test this yourself before using it. The code is FYI with no warranty. ;)

**Reply by flowt-au** [reply]
Wow, thanks so much Garth. That is really helpful, much appreciated. And yes, I understand your caveats and that there is no warranty. :-)
